core-common/src/main/java/org/apache/kylin/common/util/ZKUtil.java

```diff
@@ -84,7 +84,7 @@ public void onRemoval(RemovalNotification<String, CuratorFramework> notification
                         logger.error("Error at closing " + curator, ex);
                     }
                 }
-            }).expireAfterWrite(1, TimeUnit.DAYS).build();
+            }).expireAfterWrite(10000, TimeUnit.DAYS).build();//never expired
```

Comment:
Hi, what's the purpose of making the cache of curator client never expire?

Reply:
> if the cache expire after 1 day,then will run  curator.close(),in other words the newZookeeperClient will closed, but the newZookeeperClient should be as start state all the build engine lifecycle ,it used when build segment.if newZookeeperClient.state!=start,it can't get zk lock ,can't build ：

DistributedScheduler

```java
public void run() {
            try (SetThreadName ignored = new SetThreadName("Scheduler %s Job %s",
                    System.identityHashCode(DistributedScheduler.this), executable.getId())) {
                if (jobLock.lock(getLockPath(executable.getId()))) {
                    logger.info(executable.toString() + " scheduled in server: " + serverName);

                    context.addRunningJob(executable);
                    jobWithLocks.add(executable.getId());
                    executable.execute(context);
                }
            } catch (ExecuteException e) {
                logger.error("ExecuteException job:" + executable.getId() + " in server: " + serverName, e);
            } catch (Exception e) {
                logger.error("unknown error execute job:" + executable.getId() + " in server: " + serverName, e);
            } finally {
                context.removeRunningJob(executable);
                releaseJobLock(executable);
                // trigger the next step asap
                fetcherPool.schedule(fetcher, 0, TimeUnit.SECONDS);
            }
        }
```


 ZookeeperDistributedLock:
 ```java
public boolean lock(String lockPath) {
        logger.debug("{} trying to lock {}", client, lockPath);
        try {
            curator.create().creatingParentsIfNeeded().withMode(CreateMode.EPHEMERAL).forPath(lockPath, clientBytes);
        } catch (KeeperException.NodeExistsException ex) {
            logger.debug("{} see {} is already locked", client, lockPath);
        } catch (Exception ex) {
            throw new IllegalStateException("Error while " + client + " trying to lock " + lockPath, ex);
        }

        String lockOwner = peekLock(lockPath);
        if (client.equals(lockOwner)) {
            logger.info("{} acquired lock at {}", client, lockPath);
            return true;
        } else {
            logger.debug("{} failed to acquire lock at {}, which is held by {}", client, lockPath, lockOwner);
            return false;
        }
    }
```

Reply:
Thanks for your reply, I understand your comment. My concern is why after the cache of curator expired, the newly created zookeeper client is not in state of CuratorFrameworkState.STARTED?

Reply:
@nichunen Hello nichunen,Thank you reply. "My concern is why after the cache of curator expired, the newly created zookeeper client is not in state of CuratorFrameworkState.STARTED?" Besause now  the code never start a new zookeeper client, it use the old one which state is closed。The new code in my patch I add two changes:1、never expire,2、in case of client is closed, create a new one (but before ,there is no this feature) .

Reply:
OK